Network/Socket/Posix/Cmsg.hsc

```diff
@@ -0,0 +1,193 @@
+
+#include <sys/types.h>
+#include <sys/socket.h>
+
+import Data.ByteString.Internal
+import Foreign.ForeignPtr
+import System.IO.Unsafe (unsafeDupablePerformIO)
+import System.Posix.Types (Fd(..))
+
+import Network.Socket.Imports
+import Network.Socket.Types
+
+-- | Control message (ancillary data) including a pair of level and type.
+data Cmsg = Cmsg {
+    cmsgId   :: !CmsgId
+  , cmsgData :: !ByteString
+  } deriving (Eq, Show)
+
+----------------------------------------------------------------
+
+-- | Identifier of control message (ancillary data).
+data CmsgId = CmsgId {
+    cmsgLevel :: !CInt
+  , cmsglType :: !CInt
+  } deriving (Eq, Show)
+
+-- | The identifier for 'IPv4TTL'.
+pattern CmsgIdIPv4TTL :: CmsgId
+#if defined(darwin_HOST_OS)
+pattern CmsgIdIPv4TTL = CmsgId (#const IPPROTO_IP) (#const IP_RECVTTL)
+#else
+pattern CmsgIdIPv4TTL = CmsgId (#const IPPROTO_IP) (#const IP_TTL)
+#endif
+
+-- | The identifier for 'IPv6HopLimit'.
+pattern CmsgIdIPv6HopLimit :: CmsgId
+pattern CmsgIdIPv6HopLimit = CmsgId (#const IPPROTO_IPV6) (#const IPV6_HOPLIMIT)
+
+-- | The identifier for 'IPv4TOS'.
+pattern CmsgIdIPv4TOS :: CmsgId
+#if defined(darwin_HOST_OS)
+pattern CmsgIdIPv4TOS = CmsgId (#const IPPROTO_IP) (#const IP_RECVTOS)
+#else
+pattern CmsgIdIPv4TOS = CmsgId (#const IPPROTO_IP) (#const IP_TOS)
+#endif
+
+-- | The identifier for 'IPv6TClass'.
+pattern CmsgIdIPv6TClass :: CmsgId
+pattern CmsgIdIPv6TClass = CmsgId (#const IPPROTO_IPV6) (#const IPV6_TCLASS)
+
+-- | The identifier for 'IPv4PktInfo'.
+pattern CmsgIdIPv4PktInfo :: CmsgId
+pattern CmsgIdIPv4PktInfo = CmsgId (#const IPPROTO_IP) (#const IP_PKTINFO)
+
+-- | The identifier for 'IPv6PktInfo'.
+pattern CmsgIdIPv6PktInfo :: CmsgId
+pattern CmsgIdIPv6PktInfo = CmsgId (#const IPPROTO_IPV6) (#const IPV6_PKTINFO)
+
+-- | The identifier for 'Fd'.
+pattern CmsgIdFd :: CmsgId
+pattern CmsgIdFd = CmsgId (#const SOL_SOCKET) (#const SCM_RIGHTS)
+
+----------------------------------------------------------------
+
+-- | Locate a control message of the given type in a list of control
+--   messages. The following shows an example usage:
+--
+-- > (lookupCmsg CmsgIdIPv4TOS cmsgs >>= decodeCmsg) :: Maybe IPv4TOS
+lookupCmsg :: CmsgId -> [Cmsg] -> Maybe Cmsg
+lookupCmsg cid cmsgs = find (\cmsg -> cmsgId cmsg == cid) cmsgs
+
+-- | Filtering control message.
+filterCmsg :: CmsgId -> [Cmsg] -> [Cmsg]
+filterCmsg cid cmsgs = filter (\cmsg -> cmsgId cmsg == cid) cmsgs
+
+----------------------------------------------------------------
+
+-- | A class to encode and decode control message.
+class Storable a => ControlMessage a where
+    controlMessageId :: a -> CmsgId
+
+encodeCmsg :: ControlMessage a => a -> Cmsg
+encodeCmsg x = unsafeDupablePerformIO $ do
+    bs <- create siz $ \p0 -> do
+        let p = castPtr p0
+        poke p x
+    return $ Cmsg (controlMessageId x) bs
+  where
+    siz = sizeOf x
+
+decodeCmsg :: forall a . Storable a => Cmsg -> Maybe a
+decodeCmsg (Cmsg _ (PS fptr off len))
```

Comment:
If the `controlMessageId` function were had a `Proxy a` argument, it could be used here to check that the requested type actually has the right id for the message that's being decoded. Making the interface safer.  The current decoder ignores the message id, making the call somewhat less safe...